systems/plants/RigidBodyManipulator.cpp

```diff
@@ -1088,6 +1102,110 @@ void RigidBodyManipulator::getCMM(double* const q, double* const qd, MatrixBase<
   }
 }
 
+template <typename Scalar>
+GradientVar<Scalar, TWIST_SIZE, Eigen::Dynamic> RigidBodyManipulator::centroidalMomentumMatrix(int gradient_order)
+{
+  if (!use_new_kinsol)
+    throw std::runtime_error("method requires new kinsol format");
+
+  if (gradient_order > 1)
+    throw std::runtime_error("only first order gradient is available");
+
+  int nq = num_dof;
+  for (int i = 0; i < num_bodies; i++) {
+    Gradient<Isometry3d::MatrixType, Eigen::Dynamic>::type* dTdq = nullptr;
+    if (gradient_order > 0)
+      dTdq = &(bodies[i]->dTdq_new);
+
+    auto inertia_world = transformSpatialInertia(bodies[i]->T_new, dTdq, bodies[i]->I);
+    Ic_new[i] = inertia_world.value();
+    if (inertia_world.hasGradient()) {
+      dIc_new[i] = inertia_world.gradient().value();
+    }
+  }
+
+  for (int i = num_bodies - 1; i >= 0; i--) {
+    if (bodies[i]->hasParent()) {
+      Ic_new[bodies[i]->parent] += Ic_new[i];
+      if (gradient_order > 0) {
+        dIc_new[bodies[i]->parent] += dIc_new[i];
+      }
+    }
+  }
+
+  auto com = centerOfMass<Scalar>(gradient_order);
+  int nv = num_dof; // FLOATINGBASE TODO: assumes nq = nv
+  GradientVar<Scalar, TWIST_SIZE, Eigen::Dynamic> ret(TWIST_SIZE, nv, nq, gradient_order);
+  int gradient_row_start = 0;
+  for (int i = 0; i < num_bodies; i++) {
+    RigidBody& body = *bodies[i];
+    if (body.hasParent()) {
+      int nv_joint = body.getJoint().getNumVelocities();
+
+      ret.value().template middleCols(body.dofnum, nv_joint).noalias() = Ic_new[i] * body.J; // FLOATINGBASE TODO: assumes nq = nv
+
+      if (gradient_order > 0) {
+        ret.gradient().value().template middleRows(gradient_row_start, TWIST_SIZE * nv_joint) = matGradMultMat(Ic_new[i], body.J, dIc_new[i], body.dJdq);
+        gradient_row_start += TWIST_SIZE * nv_joint;
+      }
+    }
+  }
+
+  // TODO: could exploit structure of T better in this part
+  Eigen::Transform<Scalar, SPACE_DIMENSION, Eigen::Isometry> T(Translation<Scalar, SPACE_DIMENSION>(-com.value()));
+
+  if (gradient_order > 0) {
+    Eigen::Matrix<double, HOMOGENEOUS_TRANSFORM_SIZE, Eigen::Dynamic> dtransform_world_to_com(HOMOGENEOUS_TRANSFORM_SIZE, nq);
+    dtransform_world_to_com.setZero();
+    setSubMatrixGradient<Eigen::Dynamic>(dtransform_world_to_com, com.gradient().value(), intRange<3>(0), intRange<1>(3), SPACE_DIMENSION + 1);
+    ret.gradient().value() = dTransformAdjointTranspose(T.inverse(), ret.value(), dtransform_world_to_com, ret.gradient().value());
```

Comment:
Should modify dTransformAdjoint and turn it into dTransformSpatialForce for more convenience and efficiency.